src/peft/utils/other.py

```diff
@@ -481,3 +481,80 @@ def id_tensor_storage(tensor: torch.Tensor) -> Tuple[torch.device, int, int]:
         unique_id = storage_ptr(tensor)
 
     return tensor.device, unique_id, storage_size(tensor)
+
+
+def separate_pad_tokens(
+    input_ids: torch.Tensor, inputs_embeds: torch.Tensor, attention_mask: torch.Tensor, labels: torch.Tensor
+):
+    """
+    Separates out the tokens corresponding to padding in input_ids, inputs_embeds, attention_mask and labels Returns:
+    - input_ids: input_ids with padding tokens removed
+    - inputs_embeds: inputs_embeds with padding tokens removed
+    - attention_mask: attention_mask with padding tokens removed
+    - labels: labels with padding tokens removed
+    - pad_els: a tuple (pad_token_input_ids, pad_token_inputs_embeds, pad_token_attention_mask, pad_token_labels)
+      containing the padding token elements
+    """
+    if input_ids is None and inputs_embeds is None:
+        raise ValueError("You have to provide either input_ids or inputs_embeds")
+    ndim = input_ids.ndim if input_ids is not None else inputs_embeds.ndim - 1
+    if not ndim == 2:
+        raise ValueError("input_ids must be a 2D tensor")
+    # find the first index of a non-padding token
+    if attention_mask is not None:
+        first_non_padding_tok_ind = torch.argmax(attention_mask, dim=1)
```

Comment:
Clever. I guess it can theoretically fail if we have a row with all zeros (would return 0 for that row), but that should not happen, right? Maybe an alternative is to sum over dim=1 and use the sum as index from the end.

Reply:
Yes many cornes cases left to handle! This is an interesting one: If the attention_mask is all zeros, it's currently a valid input for HF models. I'm trying to think what this means. This is surely a malformed input, because the attention weights become zeroed out in every transformer block. I think for now I'll have consistency and change this code to make `first_non_padding_tok_ind` to be `attention_mask.shape[1]` i.e length of the sequence. I can do this simply by detecting if the `first_non_padding_tok_ind` has a zero entry. For now, I'll also enforce that input `attention_mask` should only have 0 and 1 entries, with a check in `separate_pad_tokens`. AFAIK `transformers` allows all kinds of values but ignores it internally. 

Also, sum over dim=1 and using that as the index from the end can break sometimes - there can be a corner case where there are padding tokens added at the right end by the user and we end up ignoring some tokens.

Reply:
Thanks for the detailed analysis. It is unfortunate that we have to rely on what comes down to implementation details of transformers, but I don't see a good way to make progress without making some simplifying assumptions.

> Also, sum over dim=1 and using that as the index from the end can break sometimes - there can be a corner case where there are padding tokens added at the right end by the user and we end up ignoring some tokens.

Ah yes, I didn't think of that one.

In general, I think similar kinds of problems come up again and again, but there really doesn't seem to be any solution to simplify it. PyTorch is working on adding `nested_tensor`, which seems to go in this direction, but unfortunately this feature is experimental and also would currently not support what we need here.